internal/pool/conn.go

```diff
@@ -93,3 +80,22 @@ func (cn *Conn) WithWriter(timeout time.Duration, fn func(wr *proto.Writer) erro
 func (cn *Conn) Close() error {
 	return cn.netConn.Close()
 }
+
+func (cn *Conn) deadline(ctx context.Context, timeout time.Duration) time.Time {
```

Comment:
This should probably return the lesser of the context deadline and the timeout deadline, if both are set.

Reply:
That is debatable. I personally think that this is  low-level API and it is up for user to do whatever makes sense for them.

Reply:
It's the way the timeouts and deadlines are composed in the context package. Skimming through the standard library a bit, it's how the net Dialer's DialContext method works. I think it would would be less surprising to users.

context.WithTimeout and context.WithDeadline: https://godoc.org/context#WithDeadline
net.Dialer.DialContext: https://golang.org/src/net/dial.go#L105

Reply:
Thanks, that is a good point though dialer has all sorts of knobs (Timeout, Deadline, and context) so chosing the earliest deadline makes sense there (or at least they must chose between 3).

But in go-redis I *think* situation is slightly different because Options.Read/WriteTimeout exist on different level/layer than context. First is an option for all net read/writes and second is passed explicitly as an arg and is likely different for every command.

And using context's deadline is just more flexible - sometimes you may need to use higher than default timeout.

Reply:
After some consideration I think you are right and it makes sense to copy the default approach here - https://github.com/go-redis/redis/pull/1051